### DISCUSSION TITLE
Gas showing negative hourly values

### DISCUSSION THREAD

**Discussion by jonathanmist** [BottlecapDave/HomeAssistant-OctopusEnergy#1121] [Q&A]
For the last two days the gas chart on the energy dashboard has had 2-3 negative consumption values spread through the day, each followed by a positive spike. The daily total is correct compared to the Octopus app. I'm not sure where to start debugging. Couldn't find other reports of this, so thought I'd check and see if anyone else has the same issue. 
![image](https://github.com/user-attachments/assets/7527ff76-43ea-4542-86ee-68623cfcb6c7)

**Comment by BottlecapDave** [comment]
Hello and sorry you're seeing issues. Are you using the current accumulative consumption or previous accumulative consumption sensors in the energy dashboard?

**Reply by jonathanmist** [reply]
Hi, I'm using "Current Accumulative Consumption Gas"

**Reply by BottlecapDave** [reply]
The energy dashboard takes the delta from the previous value in the sensor to work out what the values are for each hourly block. The sensor itself retrieves the 30 minutes breakdown from the API and the value is all 30 minute breakdowns added together.

I've always put this down to the API reconciling and correcting the data over the course of the day. However, if you could look at the states/attributes for the sensor over the course of the problematic day, this should give an indication if there's a problem with the internal calculation (e.g. floating point error). This can be done by using the SQL adding and using the following query (change kWh to m3 if that's your native unit)

```
SELECT  "states_meta".entity_id, "states".state,  "state_attributes".shared_attrs, DATETIME("states".last_updated_ts, 'unixepoch') AS last_updated
FROM "states_meta"
JOIN "states" ON "states".metadata_id = "states_meta".metadata_id
JOIN "state_attributes" ON "state_attributes".attributes_id = "states".attributes_id 
WHERE "states_meta".entity_id LIKE '%current_accumulative_consumption_kwh%'
AND DATETIME("states".last_updated_ts, 'unixepoch') >= '2024-12-12 00:00:00Z'
AND DATETIME("states".last_updated_ts, 'unixepoch') <= '2024-12-13 10:00:00Z'
```

**Reply by jonathanmist** [reply]
Thanks, I'm not quite sure I follow the calculations and which data is from the API vs calculated/stored, but I can see two symptoms:

1) Spurious negative "consumption" values, which I guess is the API returning duff data? They are consistently repeated in each subsequent "shared_attrs" array.

2) Earlier "state" values get repeated later. Two examples highlighted in the attached from the 12th. Maybe exiting the summation loop early somehow?
<img width="930" alt="Screenshot 2024-12-18 at 14 35 34" src="https://github.com/user-attachments/assets/9d75231a-8d44-42f3-8bcf-89dfa797e1df" />

Do you need the JSON?

**Reply by BottlecapDave** [reply]
The shared attributes is what I'm interested in, as this will have the breakdown of data come from OE on every state change. The negative records should show why it thought the state was negative.
